src/Microsoft.TemplateEngine.Cli/TemplateInvocationCoordinator.cs

```diff
@@ -0,0 +1,83 @@
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.TemplateEngine.Abstractions;
+using Microsoft.TemplateEngine.Abstractions.TemplateUpdates;
+using Microsoft.TemplateEngine.Cli.CommandParsing;
+using Microsoft.TemplateEngine.Edge.Settings;
+using Microsoft.TemplateEngine.Edge.Template;
+using Microsoft.TemplateSearch.Common.TemplateUpdate;
+
+namespace Microsoft.TemplateEngine.Cli
+{
+    internal class TemplateInvocationCoordinator
+    {
+        private readonly SettingsLoader _settingsLoader;
+        private readonly IEngineEnvironmentSettings _environment;
+        private readonly INewCommandInput _commandInput;
+        private readonly ITelemetryLogger _telemetryLogger;
+        private readonly string _commandName;
+        private readonly Func<string> _inputGetter;
+        private readonly New3Callbacks _callbacks;
+
+        public TemplateInvocationCoordinator(SettingsLoader settingsLoader, INewCommandInput commandInput, ITelemetryLogger telemetryLogger,  string commandName, Func<string> inputGetter, New3Callbacks callbacks)
+        {
+            _settingsLoader = settingsLoader;
+            _environment = _settingsLoader.EnvironmentSettings;
+            _commandInput = commandInput;
+            _telemetryLogger = telemetryLogger;
+            _commandName = commandName;
+            _inputGetter = inputGetter;
+            _callbacks = callbacks;
+        }
+
+        public async Task<CreationResultStatus> CoordinateInvocationOrAcquisitionAsync(ITemplateMatchInfo templateToInvoke)
```

Comment:
consider adding optional cancellation token params to your async methods & passing the tokens up the stack to other async calls

Reply:
this was extracted to task https://github.com/dotnet/templating/issues/2716 as it is not in scope of this change.